src/management-system-v2/lib/data/file-manager/file-manager.ts

```diff
@@ -0,0 +1,206 @@
+'use server';
+
+import { Storage } from '@google-cloud/storage';
+import fse from 'fs-extra';
+import path from 'path';
+import envPaths from 'env-paths';
+import { LRUCache } from 'lru-cache';
+
+// Constants
+const MAX_CONTENT_LENGTH = 10 * 1024 * 1024; // 10 MB
+const DEPLOYMENT_ENV = process.env.NEXT_PUBLIC_DEPLOYMENT_ENV as 'cloud' | 'local';
+const BUCKET_NAME = process.env.GOOGLE_CLOUD_BUCKET_NAME || '';
+const cache = new LRUCache<string, Buffer>({
+  maxSize: 100,
+  ttl: 60 * 60 * 1000, // 1 hour TTL
+  sizeCalculation: () => 1,
+});
+
+const getLocalStorageBasePath = (): string => {
+  let appDir = envPaths('proceed-management-system').config;
+  appDir = appDir.slice(0, appDir.search('-nodejs'));
+  return process.env.NODE_ENV === 'development' ? path.join(appDir, 'development') : appDir;
+};
+
+const LOCAL_STORAGE_BASE = getLocalStorageBasePath();
+let storage: Storage | null = null;
+let bucket: any = null;
+
+if (DEPLOYMENT_ENV === 'cloud') {
+  storage = new Storage({ keyFilename: process.env.GCP_KEY_PATH });
+  bucket = storage.bucket(BUCKET_NAME);
+}
+
+// Helper functions
+const setCors = async (bucket: any) => {
+  await bucket.setCorsConfiguration([
+    {
+      maxAgeSeconds: 3600,
+      method: ['GET', 'PUT'],
+      origin: ['*'], // Adjust trusted origin for production
+      responseHeader: ['content-type', 'x-goog-content-length-range'],
+    },
+  ]);
+};
+
+const ensureBucketExists = () => {
+  if (!bucket) throw new Error('Storage bucket not initialized');
+};
+
+const saveLocalFile = async (filePath: string, fileContent: Buffer) => {
+  const fullPath = path.join(LOCAL_STORAGE_BASE, filePath);
+  await fse.ensureDir(path.dirname(fullPath));
+  await fse.writeFile(fullPath, fileContent);
+  cache.set(filePath, fileContent);
+};
+
+// Core Functions
+export const saveFile = async (
+  filePath: string,
+  mimeType: string,
+  fileContent?: Buffer | Uint8Array | Blob | string,
+  usePresignedUrl: boolean = true,
+): Promise<{ presignedUrl: string | null; status: boolean }> => {
+  let presignedUrl: string | null = null;
+  let status = false;
+
+  try {
+    if (DEPLOYMENT_ENV === 'cloud') {
+      ensureBucketExists();
+      const file = bucket.file(filePath);
+
+      if (usePresignedUrl) {
+        // Generate a presigned URL for file upload
+        [presignedUrl] = await file.getSignedUrl({
+          version: 'v4',
+          action: 'write',
+          expires: Date.now() + 15 * 60 * 1000, // 15 minutes
```

Comment:
its for upload start, once the link is used, the GCP generates a session URI for upload. 

found this on google documentation:

_'because after the request to [initiate the upload](https://cloud.google.com/storage/docs/performing-resumable-uploads#initiate-session), subsequent PUT requests to upload the object data use a session URI, which acts as an authentication token. This means that PUT requests don't use any signed URLs.'_